### PR TITLE
Refactor util/mkdef.pl for clearer separation of functionality

### PR DESCRIPTION
**Pull request by levitte**
Main modifications (from commits messages)
-------------------

Move the .num updating functionality to util/mknum.pl.
Rewrite util/mkdef.pl to create .def / .map / .opt files exclusively,
using the separate ordinals reading module. 
Adapt the build files.
Adapt the symbol presence test.

Rewrite util/mknum.pl to become cleaner, and to use the separate
generic C header parsing module, as well as the separate ordinals
manipulation module.
Adapt the build files.

Additional features (from commit messages)
-----------------------

- Add a perl module that loads and manipulates ordinals files
  (OpenSSL::Ordinals)

  This means adding the capability to add new items, to invalidate and
  revalidate all the items, and to update the file it came from, as well
  as the possibility to create new items from other data than a line
  from said file.
    
- Add a perl module to parse through C headers

  OpenSSL::ParseC is a module that parses through a C header file and
  returns a list with information on what it found.  Currently, the
  information it returns covers function and variable declarations,
  macro definitions, struct declarations/definitions and typedef
  definitions.

### CODE REVIEW

**Comment by levitte** [comment]
The goal is to parametrize these scripts instead of having so much hard coded stuff.  That will allow us much more flexibility in handling more libraries, modules, etc

File: util/mknum.pl [hunk 1]
```diff
@@ -0,0 +1,1357 @@
...
+#   used.  For the positive items, if all of them are false in the environment,
+#   the corresponding symbol can't be used.  Any combination of positive and
+#   negative items are possible, and of course leave room for some redundancy.
+# - "kind" is "FUNCTION" or "VARIABLE".  The meaning of that is obvious.
+# - "algorithms" is a comma-separated list of algorithm names.  This helps
+#   exclude symbols that are part of an algorithm that some user wants to
+#   exclude.
+#
+
+use lib ".";
```
Comment (richsalz) [inline util/mknum.pl]: Shouldn't this have "use strict; use warnings;" and pass cleanly?

Comment (levitte) [reply util/mknum.pl]: I'm still working on this script, my plan is to refactor it pretty brutally.  This is a mere start for a proof of concept.  See "WIP" :wink:

Comment (richsalz) [reply util/mknum.pl]: just confirming your intent. :)

**Comment by levitte** [comment]
(note that this currently fails in the CIs because #7331 needs being merged first...  these scripts don't much like duplicate declarations, and the question is if we should make them, or if we should see this as an opportunity to keep cleaner)

**Comment by richsalz** [comment]
Avoiding the duplicates is a *great* idea.

**Comment by levitte** [comment]
I've made a gazillion tests and am reasonably satisfied with the result, so it's time this gets reviewed.

This is a step toward being able to build modules fairly hassle free with the same set of scripts.  I have a PR coming up where engines are getting an export table with no symbol versions, to demonstrate what can be done.

**Comment by levitte** [comment]
Note that there's an expected failure in Travis because `util/mknum.pl` wants to add `SSL_version_str`, see #7340...

**Comment by levitte** [comment]
I just rebased on top of fresh master with #7340 merged.  I think the CI runs will be error free now...  or well, I hope.

**Comment by levitte** [comment]
CIs are all happy.

**Comment by t-j-h** [comment]
I've been waiting on the CIs as there are a lot of changes there.
Can you add in the normal OpenSSL copyright on those files ... they should be in place.

**Comment by t-j-h** [comment]
BTW it is good to see a pile of rather old perl code get replaced ...

**Comment by levitte** [comment]
The only file that was added in this PR that lacked the copyright boilerplate was `util/perl/OpenSSL/ParseC.pm`.  I just pushed a fixup that adds it.
Already existing files lacking a boilerplate should be taken care of with a separate cleanup PR.

**Comment by levitte** [comment]
(I'll probably get to refactor ParseC soon...  I can see reasons to have it parse through C code as well...)

**Review by t-j-h — APPROVED** [review]

**Comment by levitte** [comment]
Thank you!  Merged.

30699aa1943b10b265c52334d9f582c04c4eccba Refactor util/mknum.pl for clearer separation of functionality
15ba109631254b98d3ef2611a1765e75442314a6 Add code to manipulate the items in OpenSSL::Ordinals
d73c44404d470424aa58e85fe38b97351f112bc1 A perl module to parse through C headers
ab1e5495e45708f43f71e1f9e8872066dc8177c0 Move ZLIB from 'platforms' to 'features'
8effd8fa67689e1d4318cfda21604eef428a37cf Refactor util/mkdef.pl for clearer separation of functionality
91a99748d328164ab043cf7bc3da8e45ec0de497 Add a perl module that deals with ordinals files

**Pull request closed without merging**